utils/src/lib.rs

```diff
@@ -223,55 +223,45 @@ pub fn etherscan_api_key() -> eyre::Result<String> {
-pub fn abi_to_solidity(contract_abi: Abi, contract_name: String) -> Result<String> {
-    let mut out: Vec<String> = vec![];
-    let functions_iterator = contract_abi.functions();
-    out.push(String::from("interface "));
-    out.push(contract_name);
-    out.push(String::from(" { \n"));
-    for function in functions_iterator {
-        out.push(String::from("     "));
-        out.push(function.name.clone());
-        out.push(String::from("("));
-        if function.inputs.len() != 0 {
-            for input in function.inputs.clone() {
-                out.push(String::from(" "));
-                out.push(input.kind.to_string());
-                out.push(String::from(" "));
-                out.push(input.name.to_string());
-                out.push(String::from(", "));
-            }
-            out.pop();
-        }
-        out.push(String::from(") "));
-        match function.state_mutability {
-            abi::StateMutability::Pure => {
-                out.push(String::from("pure "));
-            }
-            abi::StateMutability::View => {
-                out.push(String::from("view "));
-            }
-            abi::StateMutability::Payable => {
-                out.push(String::from("payable "));
+pub fn abi_to_solidity(contract_abi: &Abi, contract_name: &str) -> Result<String> {
+    let functions_iterator = contract_abi.functions().into_iter();
+    let functions = functions_iterator
+        .map(|function| {
+            let inputs = function
+                .inputs
+                .iter()
+                .map(|input| format!("{} {}", input.kind, input.name))
+                .collect::<Vec<String>>()
+                .join(", ");
+            let outputs = function
+                .outputs
+                .iter()
+                .map(|output| format!("{} {}", output.kind, output.name))
+                .collect::<Vec<String>>()
+                .join(", ");
+            let mut mutability: &str = "";
+            match function.state_mutability {
+                abi::StateMutability::Pure => {
+                    mutability = "pure";
+                }
+                abi::StateMutability::View => {
+                    mutability = "view";
+                }
+                abi::StateMutability::Payable => {
+                    mutability = "payable";
+                }
+                _ => {}
             }
-            _ => {}
-        }
-        if function.outputs.len() != 0 {
-            out.push(String::from("returns"));
-            out.push(String::from("("));
-            for output in function.outputs.clone() {
-                out.push(String::from(" "));
-                out.push(output.kind.to_string());
-                out.push(String::from(" "));
-                out.push(output.name.to_string());
-                out.push(String::from(", "));
+            if outputs.len() == 0 {
+                format!("\n     {}({}) {} external;", function.name, inputs, mutability)
```

Comment:
Think you forgot adding `function` before the function name?

Reply:
lol yy